source/io/module_javaScriptInvoke.js

```diff
@@ -131,156 +138,123 @@ var assignJavaScriptInvoke;
+                },
+
+                visitString: function (valueRef) {
+                    return Module.eggShell.readString(valueRef);
+                },
+
+                visitArray: function (valueRef) {
+                    return Module.eggShell.readTypedArray(valueRef);
+                },
+
+                visitJSObjectRefnum: function (valueRef) {
+                    return Module.eggShell.readJavaScriptRefNum(valueRef);
                 }
-            }
+            };
         };
 
-        var createJavaScriptParametersArray = function (parametersPointer, parametersCount) {
-            var parameters = new Array(parametersCount);
-            for (var index = 0; index < parametersCount; index += 1) {
-                var typeName = getParameterTypeString(parametersPointer, index);
-                var parameterPointer = JavaScriptInvoke_GetParameterPointer(parametersPointer, index);
-                if (typeName === 'Array') {
-                    typeName += getArrayElementTypeString(parameterPointer);
+        var jsInvokePeeker = createJavaScriptInvokePeeker();
+        var peekValueRef = function (valueRef) {
+            return Module.eggShell.reflectOnValueRef(jsInvokePeeker, valueRef);
+        };
+
+        var createJavaScriptInvokePoker = function () {
+            var visitNumeric = function (valueRef, data) {
+                if (typeof data.userValue !== 'number') {
+                    mergeNewError(data.errorValueRef, data.functionName, ERRORS.kNITypeMismatchForReturnTypeInJavaScriptInvoke);
+                    return;
                 }
+                Module.eggShell.writeDouble(valueRef, data.userValue);
+            };
 
-                var parameterValue = undefined;
-                var readFunction = typeFunctions[typeName].reader;
-                if (readFunction === undefined) {
-                    throw new Error(' Unsupported type = ' + typeName + ' for parameter with index = ' + index);
-                } else {
-                    parameterValue = readFunction(parameterPointer);
+            return {
+                visitInt8: visitNumeric,
+                visitInt16: visitNumeric,
+                visitInt32: visitNumeric,
+                visitUInt8: visitNumeric,
+                visitUInt16: visitNumeric,
+                visitUInt32: visitNumeric,
+                visitSingle: visitNumeric,
+                visitDouble: visitNumeric,
+                visitEnum8: visitNumeric,
+                visitEnum16: visitNumeric,
+                visitEnum32: visitNumeric,
+
+                visitBoolean: function (valueRef, data) {
+                    if (typeof data.userValue !== 'boolean') {
+                        mergeNewError(data.errorValueRef, data.functionName, ERRORS.kNITypeMismatchForReturnTypeInJavaScriptInvoke);
+                        return;
+                    }
+                    Module.eggShell.writeDouble(valueRef, data.userValue ? 1 : 0);
```

Comment:
These writes should be wrapped in try catch to be able to report the "Unable to set return value" errors

Reply:
oops. Moved the try catch around reflectOnValueRef in pokeValueRef.